core/psh/sysexec.c

```diff
@@ -25,6 +27,63 @@ void psh_sysexecinfo(void)
 }
 
 
+static int psh_sysexec_wldargmatch(char *wld, int argc, char **argv, char wldend, int argmaxlen)
+{
+	int len = 0, ii;
+	char *currc = wld, *nextc, *cmdend;
+
+	cmdend = strchr(wld, wldend);
+	currc = wld;
+	for (ii = 0; ii < argc; ii++) {
+
+		if ((nextc = strchr(currc, ' ')) == NULL || nextc > cmdend)
+			nextc = cmdend;
+		if ((nextc == cmdend) ^ (ii == (argc - 1))) /* exits if: last argc XOR last wldcrd argument */
+			return -1;
+
+		len = nextc - currc;
+		if (len > argmaxlen || len == 0 || strlen(argv[ii]) != len || memcmp(argv[ii], currc, len) != 0)
+			return -1;
+		if (*nextc != wldend && nextc != NULL)
+			currc = nextc + 1;
+	}
+	
+	return 1;
+}
+
+
+static int psh_sysexec_checkcommand(int argc, char **argv)
+{
+	int maxcdmlen = 80, acc = -1;
+	char wldbuff[maxcdmlen], *wldlist = PSH_WILDCARDS, *curr = wldlist, *nxtcomma;
```

Comment:
`wldlist` may be defined as `static const`, as `PSH_WILDCARDS` is a string constant.

Reply:
Why is this variable defined in the first place? It's only used to be assigned to `curr`. I guess to work around valid warning - that you assign `const` string to `char` variable - so you access `const` string using pointer to mutable string!

`curr[] = PSH_WILDCARDS` should be declared instead (or something what would not access const string as not const string).

Reply:
I would revise the code for use of `const char *` pointers  instead of `char *` referencing the string literal underneath `PSH_WILDCARDS` define (which string may go to `rodata` as it is only read in the code), instead of coping it with flex `[]` (without `const`) onto the stack that makes the string mutable. The `PSH_WILDCARDS` as string `""` literal is never changed in the code, only readed, neither `strchr()`, `strlen()`, nor `memcpy()` modifies its content, moreover I'll rewrite function prototype to:
```C
static int psh_sysexec_wldargmatch(const char *wld, int argc, char **argv, char wldend, int argmaxlen)
```
also adjusting `char *currc = wld, *nextc, *cmdend` declarations to `const char *currc = wld, *nextc, *cmdend` inside the aforementioned function body.

Both functions `psh_sysexec_checkcommand()` and `psh_sysexec_wldargmatch()` do not modify the content of `PSH_WILDCARDS` so it could be in `rodata`, they only work on pointers and read the content and compare. It's perfectly fine to assign `const char *`  <-- a <-- `char *` (see direction of arrows), but not the other way. That's why @agkaminski wrote about ''working around the valid warning".

That way `PSH_WILDCARDS` will be put in `rodata`, so `static const char wldlist[] = PSH_WILDCARDS`, referenced with `const char *curr = wldlist;` `const char *nxtcomma` as pointers to the read-only string, but `char wldbuff[80]` should be as is, and please use `sizeof(wldbuff)` operator instead of variable `maxcdmlen` (current spelling), You don't need variable here.

Reply:
Thank you both for elaborating. As far as use of such three variables was not intended dodge of the warning, the fact of accessing, what is to be, const string as non const string is valid mistake. I will stick to three variables approach to clearly distinguish what I am pointing at (`wldlist`) and what I am pointing with (`curr`,`nxtcomma`) with definitions proposed by @gerard5:
`static const char wldlist[] = PSH_WILDCARDS`
`const char *curr = wldlist, *nxtcomma;`
also with propagating _const-ness_ of these variables in `psh_sysexec_wldargmatch`:
`const char *currc = wld, *nextc, *cmdend`

Also `maxcdmlen/maxcmdlen` is removed in favor of `sizeof`.
Thanks